common/src/main/java/org/apache/uniffle/common/netty/protocol/Encoders.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.common.netty.protocol;
+
+import java.util.List;
+
+import io.netty.buffer.ByteBuf;
+
+import org.apache.uniffle.common.ShuffleBlockInfo;
+import org.apache.uniffle.common.ShuffleServerInfo;
+import org.apache.uniffle.common.util.ByteBufUtils;
+
+public class Encoders {
+  public static void encodeShuffleServerInfo(ShuffleServerInfo shuffleServerInfo, ByteBuf byteBuf) {
+    ByteBufUtils.writeLengthAndString(byteBuf, shuffleServerInfo.getId());
+    ByteBufUtils.writeLengthAndString(byteBuf, shuffleServerInfo.getHost());
+    byteBuf.writeInt(shuffleServerInfo.getGrpcPort());
+    byteBuf.writeInt(shuffleServerInfo.getNettyPort());
+  }
+
+  public static void encodeShuffleBlockInfo(ShuffleBlockInfo shuffleBlockInfo, ByteBuf byteBuf) {
+    byteBuf.writeInt(shuffleBlockInfo.getPartitionId());
+    byteBuf.writeLong(shuffleBlockInfo.getBlockId());
+    byteBuf.writeInt(shuffleBlockInfo.getLength());
+    byteBuf.writeInt(shuffleBlockInfo.getShuffleId());
+    byteBuf.writeLong(shuffleBlockInfo.getCrc());
+    byteBuf.writeLong(shuffleBlockInfo.getTaskAttemptId());
+    ByteBufUtils.copyByteBuf(shuffleBlockInfo.getData(), byteBuf);
```

Comment:
Do we need to release memory after copying here?
```
shuffleBlockInfo.getData().release().
```

Reply:
> Do we need to release memory after copying here?
> 
> ```
> shuffleBlockInfo.getData().release().
> ```

If ByteBuf uses heap memory, I think there is no need to manually call release and this part of memory will be reclaimed during gc. Secondly, there may be retries. The current encode does not mean that the request must be  successful.

Reply:
> Could we use CompositeByteBuf to avoid data copy here?

It doesn’t seem to work. The byteBuf here comes from MessageEncoder. I checked the source code of Spark. How does it handle it? Spark will divide the Message into Header and body. The body is usually a ByteBuf and implements a MessageWithHeader (link: https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/protocol/MessageWithHeader.java), but also cannot avoid copy, see line 123: `writtenBody = copyByteBuf((ByteBuf) body, target);`

Reply:
When to use copy? When to use readSlice/CompositeByteBuf?
When we need to pass the content of ByteBuf A to ByteBuf B, `if we can manage ByteBuf B, then we can avoid copying` (such as through the readSlice method), When B will not be accessed anywhere, and its content will not passed to another ByteBuf, we can call the release method of B, and A will also be released at this time

Reply:
If it's difficult to implement, we can add a `todo` here.

Reply:
> You can see that Spark copy `Bytebuf` to `Channel`. It's ok. We copy `ByteBuf` to `ByteBuf` here. This should be optimized.


We will implement similar logic when supporting zero-copy.

Reply:
> > Do we need to release memory after copying here?
> > ```
> > shuffleBlockInfo.getData().release().
> > ```
> 
> If ByteBuf uses heap memory, I think there is no need to manually call release and this part of memory will be reclaimed during gc. Secondly, there may be retries. The current encode does not mean that the request must be successful.

I think we better call `release()` method here, see https://netty.io/wiki/reference-counted-objects.html.